### DISCUSSION TITLE
Blade directive sets all expressions to string.

### DISCUSSION THREAD

**Discussion by lucasantoniooficial** [laravel/framework#45026] [Ideas]
I see this as a limitation when we need to create a blade directive, in the documentation example it makes sense in terms of how we can have a datetime directive.

But I found a limitation in case we don't want to send a Datetime instance but a string with date.

Within my example I created a datetime directive and with it I send a string that is a date.

`@datetime("2022-11-18")`

If inside the directive function I do `dd($expression)`, it will return `""2022-11-18""`

I've used `Blade::if` and didn't have this problem with it, shouldn't we be able to do complex things with the Blade directive?

**Comment by arbazargani** [comment]
please provide more details, like your directive definition in AppServiceProvier.

**Comment by ryangjchandler** [comment]
Directives registered via `Blade::directive()` behave differently from the ones registered with `Blade::if()`. Blade's compiler is essentially a giant `preg_replace()`.

When it encounters your custom directives, i.e. `@datetime()`, it will invoke the function that you register and replace the directive code with the string returned from your directive.

**Comment by isDOSe** [comment]
Another example

 `Blade::directive('statusBadge', function ($id) {
            dd($id); });`

if I try this `@statusBadge($lastLoan->status_id)` in my views it will show "$lastLoan->status_id" not the number :(

Sorry for my english 😌

**Comment by Yinci** [comment]
For the people who still come across this (because I did), listen up:

It isn't actually an issue. It's how the Blade directive works. You pass an expression, and you need to have the expression within the returned string, because that's how it is evaluated. So, say you have:

```php
$var = 'Hello';

@world($var); // Hello World

Blade::directive('world', function ($expression) {
    dump($expression); // $var
    return '<?php echo $expression World'?>
});
```

Don't forget to clear view cache after changing the directive.
